onnxscript/analysis.py

```diff
@@ -51,31 +49,39 @@ def get_id(e):
     return {get_id(lhs)}
 
 
-def defs(stmt: ast.stmt) -> Set[str]:
-    """Return the set of all variables that may be defined (assigned to) in an
-    execution of input stmt.
+def assigned_vars(
+    stmt: ast.stmt | list[ast.stmt], formatter: sourceinfo.Formatter
+) -> Set[str]:
+    """Return the set of all variables that may be assigned to in an execution of input stmt
+    or sequence of statements.
     """
 
-    def block_defs(block: Sequence[ast.stmt]) -> Set[str]:
+    def assigned_in_block(block: Sequence[ast.stmt]) -> Set[str]:
         result: set[Any] = set()
         for s in block:
-            result = result | defs(s)
+            result = result | assigned_vars(s, formatter)
         return result
 
     if isinstance(stmt, ast.Assign):
-        return local_defs(stmt.targets[0])
+        return lhs_vars(stmt.targets[0])
     if isinstance(stmt, ast.AnnAssign):
-        return local_defs(stmt.target)
+        return lhs_vars(stmt.target)
     if isinstance(stmt, ast.Return):
         return set()
     if isinstance(stmt, ast.If):
-        return block_defs(stmt.body) | block_defs(stmt.orelse)
+        return assigned_in_block(stmt.body) | assigned_in_block(stmt.orelse)
+    if isinstance(stmt, ast.For):
+        return assigned_in_block(stmt.body) | {get_loop_var(stmt, formatter)}
+    if isinstance(stmt, ast.While):
+        return assigned_in_block(stmt.body)
     if isinstance(stmt, list):
-        return block_defs(stmt)
+        return assigned_in_block(stmt)
     if isinstance(stmt, ast.Break):
         return set()
     if ast_utils.is_print_call(stmt):
         return set()
+    if isinstance(stmt, list):
+        return assigned_in_block(stmt)
```

Comment:
Should we test this too?

Reply:
It is tested by every test for this function, since `parse_tree.body` is a list of statements. In fact, this extension exists primarily for enabling the unit tests.

Reply:
hmm strange. 
![image](https://github.com/microsoft/onnxscript/assets/11205048/1b96e65d-4070-4aaa-a472-48e8e55b43c2)

coverage shows only this line is not tested

Reply:
Let me print some things here to see

Reply:
Duplicate of line 77?

Reply:
Good catch, thanks!